src/amperity/gocd/secret/vault/plugin.clj

```diff
@@ -240,12 +262,14 @@
   (try
     (when-not @client
       (authenticate-client-from-inputs! client (:configuration data)))
-    (let [secrets (lookup-secrets @client (:keys data))
-          missing-keys (mapv :key (remove :value secrets))]
+    (let [{token-keys true secrets-keys false} (group-by #(str/starts-with?  % signify-token-creation-str) (:keys data))
+          secrets (lookup-secrets @client secrets-keys)
```

Comment:
i would prefer doing fewer side-actions in the `let` binding as a cleanliness thing.

I also don't think this method needs to support both actions at once, it should be one or the other.

Reply:
Could you clarify what you mean by side-actions?

Also, I think supporting both actions is just as easy as (or even simpler than) supporting one. All the lookup logic is still separated into two functions (for token lookups and secret keys) Plus, this way it has the expected behavior if a user put both in the configurations (imagine a pipeline using an environment that looks up some secrets among other useful env vars, but with a token lookup in the pipeline specific params)

Reply:
> Could you clarify what you mean by side-actions?

Side-effects happening in the course of this function executing. Not that functions in a GoCD plugin should expect to be pure, but it tends to be good practice to avoid too many side-effecting actions in let bindings if avoidable. As written, the ordering and dependency of the actions is implicit so its a little harder to read and reason about.

"Lookup keys, report and fail if missing" unit of work
```clojure
    secrets (lookup-secrets @client secrets-keys) ;; side-effects: look up secrets
    missing-keys (mapv :key (remove :value secrets))
,,,
 (if (empty? missing-keys)
   ;; return success
```

"Create and return tokens" unit of work
```clojure
    all-lookups (concat (create-tokens! @client token-keys) secrets) ;; side-effects: create tokens before checking previous for errors.
,,,
   ;; only in "no missing keys branch" return tokens
   ;; otherwise, tokens are created and thrown away
```

This interleaves execution in a way that results in token creation even if errors would be returned.

Instead, it might be more clean to force these into a threading flow to more fully separate them.

```clojure
(let [{token-keys true secrets-keys false} (group-by #(str/starts-with?  % signify-token-creation-str) (:keys data))
      secrets (lookup-secrets @client secrets-keys)]
  (if-let [missing-keys (->> (remove :value secrets)
                             (map :key)
                             (seq))]
    {:response-code    404
     :response-headers {}
     :response-headers {}
     :response-body    {:message (str "Unable to resolve key(s) " missing-keys)}}
    {:response-code    200
     :response-headers {}
     :response-headers {}
     :response-body (-> (create-tokens! @client token-keys)
                        (concat secrets)
                        (->> (mapv #(update % :value str))))}))
```

In this form the side-effects are at least more separated in execution/gated flow: token creation only happens if no secret keys were not found. 

I don't think it was your intention to create tokens even if keys are missing, so hopefully this illustrates why making let bindings too complex can be a problem.

Reply:
>  Plus, this way it has the expected behavior if a user put both in the configurations (imagine a pipeline using an environment that looks up some secrets among other useful env vars, but with a token lookup in the pipeline specific params)

Oh of course, I was thinking you were trying to support these two methods _from the same key_ but that makes sense there could be multiple keys each with different types. In my mind our internal intention is to use one or the other, but this is not necessarily true!

Reply:
> I don't think it was your intention to create tokens even if keys are missing, so hopefully this illustrates why making let bindings too complex can be a problem.

It shouldn't create tokens if keys are missing anyway, since `create-tokens!` returns a lazy seq. That being said:

> Instead, it might be more clean to force these into a threading flow to more fully separate them.

I agree, it looks cleaner to me. I'll modify the code! 👍